controllers/managedclustermodule_reconciler.go

```diff
@@ -0,0 +1,141 @@
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+	workv1 "open-cluster-management.io/api/work/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	kmmv1beta1 "github.com/kubernetes-sigs/kernel-module-management/api/v1beta1"
+	"github.com/kubernetes-sigs/kernel-module-management/internal/manifestwork"
+)
+
+// ManagedClusterModuleReconciler reconciles a ManagedClusterModule object
+type ManagedClusterModuleReconciler struct {
+	client.Client
+
+	manifestAPI manifestwork.ManifestWorkCreator
+}
+
+//+kubebuilder:rbac:groups=kmm.sigs.k8s.io,resources=managedclustermodules,verbs=get;list;watch;update;patch
+//+kubebuilder:rbac:groups=kmm.sigs.k8s.io,resources=managedclustermodules/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=work.open-cluster-management.io,resources=manifestworks,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=cluster.open-cluster-management.io,resources=managedclusters,verbs=get;list;watch
+
+func NewManagedClusterModuleReconciler(
+	client client.Client,
+	manifestAPI manifestwork.ManifestWorkCreator) *ManagedClusterModuleReconciler {
+	return &ManagedClusterModuleReconciler{
+		Client:      client,
+		manifestAPI: manifestAPI,
+	}
+}
+
+func (r *ManagedClusterModuleReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	res := ctrl.Result{}
+
+	logger := log.FromContext(ctx)
+
+	mcm, err := r.getRequestedManagedClusterModule(ctx, req.NamespacedName)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			logger.Info("ManagedClusterModule deleted")
+			return res, nil
+		}
+
+		return res, fmt.Errorf("failed to get the requested CR: %v", err)
+	}
+
+	logger.Info("Requested KMMO ManagedClusterModule")
+
+	clusters, err := r.getSelectedManagedClusters(ctx, mcm)
+	if err != nil {
+		return res, fmt.Errorf("failed to get selected clusters: %v", err)
+	}
+
+	if err := r.manifestAPI.GarbageCollect(ctx, *clusters, *mcm); err != nil {
```

Comment:
I think it is better to issue garbage collection for manifests at the end of the reconcile loop.

Reply:
@yevgeny-shnaidman thank you for the suggestion, I changed the code accordingly :+1: